### PR TITLE
Fix issues revealed by valgrind; add valgrind check on CI.

### PR DESCRIPTION
**Pull request by cqc-alec**

### CODE REVIEW

File: tket/src/Transformations/ControlledGates.cpp [hunk 1]
```diff
@@ -432,22 +432,27 @@ static Circuit lemma71(
...
+    if (!bin.contains(v) && rep.get_OpType_from_Vertex(v) == OpType::CRy) {
       Expr v_angle = rep.get_Op_ptr_from_Vertex(v)->get_params()[0];
       Circuit cry_replacement = CircPool::CRy_using_CX(v_angle);
       Subcircuit sub{rep.get_in_edges(v), rep.get_all_out_edges(v), {v}};
-      rep.substitute(cry_replacement, sub, Circuit::VertexDeletion::Yes);
+      rep.substitute(cry_replacement, sub, Circuit::VertexDeletion::No);
+      bin.insert(v);
     }
-    if (rep.get_OpType_from_Vertex(v) == OpType::CU1) {
+    if (!bin.contains(v) && rep.get_OpType_from_Vertex(v) == OpType::CU1) {
```
Comment (yao-cqc) [inline tket/src/Transformations/ControlledGates.cpp:446]: Would changing this to `else if` solve the pointer referencing problem?

Comment (cqc-alec) [reply tket/src/Transformations/ControlledGates.cpp:446]: Yes the code could definitely be better structured...

Comment (cqc-alec) [reply tket/src/Transformations/ControlledGates.cpp]: Done.

**Review by yao-cqc — APPROVED** [review]
Nice, thanks!

**Pull request merged**